File: implementations/csharp/client/tests/Specus.Client.Tests/PeerEgressRouteVectorTests.cs
using System.Text.Json;
using Specus.Client.PeerMesh;
using Specus.Protocol.PeerEgress;

namespace Specus.Client.Tests;

/// <summary>
/// Replays <c>peer-egress-routes-v1.json</c> against the .NET route planner.
/// </summary>
/// <remarks>
/// Unlike the TCP fixture, this one's expectations come from an independent reference planner in
/// <c>tools/protocol/generate_peer_egress_route_vectors.py</c> rather than from any implementation,
/// so agreeing with it is independent agreement rather than a shared mistake.
///
/// <para>The operator-visible <c>origin</c> string is asserted too. Three runtimes that installed
/// the same prefixes but described them differently would still leave an operator unable to compare
/// one machine's log with another's.</para>
/// </remarks>
public class PeerEgressRouteVectorTests
{
    private static JsonDocument ReadVector(string name)
    {
        var directory = new DirectoryInfo(AppContext.BaseDirectory);
        while (directory is not null)
        {
            var candidate = Path.Combine(directory.FullName, "protocol", "test-vectors", name);
            if (File.Exists(candidate))
            {
                return JsonDocument.Parse(File.ReadAllText(candidate));
            }
            directory = directory.Parent;
        }
        throw new FileNotFoundException($"cannot locate {name}");
    }

    private static List<PeerEgressRoute> RoutesOf(JsonElement array)
    {
        var routes = new List<PeerEgressRoute>();
        foreach (var node in array.EnumerateArray())
        {
            routes.Add(new PeerEgressRoute(
                node.GetProperty("cidr").GetString()!,
                PeerEgressRoutePlanner.KindFromWireName(node.GetProperty("kind").GetString()),
                node.GetProperty("origin").GetString()!));
        }
        return routes;
    }

    private static void AssertRoutes(string label, IReadOnlyList<PeerEgressRoute> got, JsonElement want)
    {
        var expected = RoutesOf(want);
        Assert.True(expected.Count == got.Count,
            $"{label}: planned {got.Count} routes, want {expected.Count}: {string.Join(", ", got)}");
        for (var index = 0; index < expected.Count; index++)
        {
            Assert.Equal(expected[index], got[index]);
        }
    }

    private static PeerEgressRule RuleOf(JsonElement node) => new()
    {
        Match = node.TryGetProperty("match", out var match) ? match.GetString() ?? string.Empty : string.Empty,
        Action = node.TryGetProperty("action", out var action) ? action.GetString() ?? string.Empty : string.Empty,
        EgressClientId = node.TryGetProperty("egressClientId", out var target) ? target.GetInt64() : null,
        Port = node.TryGetProperty("port", out var port) ? port.GetInt32() : null,
    };

    [Fact]
    public void RoutePlansMatchTheSharedVector()
    {
        using var vector = ReadVector("peer-egress-routes-v1.json");
        var meshCidr = vector.RootElement.GetProperty("meshCidr").GetString();
        var planCases = vector.RootElement.GetProperty("planCases");
        Assert.True(planCases.GetArrayLength() > 0, "routes vector carried no plan cases");

        foreach (var testCase in planCases.EnumerateArray())
        {
            var name = testCase.GetProperty("name").GetString()!;
            var rules = testCase.GetProperty("rules").EnumerateArray().Select(RuleOf).ToList();
            var bypass = testCase.GetProperty("bypass").EnumerateArray()
                .Select(node => node.GetString() ?? string.Empty).ToList();

            var plan = PeerEgressRoutePlanner.Plan(rules, bypass, meshCidr);
            AssertRoutes(name, plan.Routes, testCase.GetProperty("expect").GetProperty("routes"));

            // Asserted alongside the routes because a planner that silently dropped a bad rule
            // would produce exactly the same route list.
            var wantRefused = testCase.GetProperty("expect").GetProperty("refused");
            Assert.True(wantRefused.GetArrayLength() == plan.Refused.Count,
                $"{name}: refused {plan.Refused.Count} rules, want {wantRefused.GetArrayLength()}");
            var index = 0;
            foreach (var want in wantRefused.EnumerateArray())
            {
                var got = plan.Refused[index];
                Assert.Equal(want.GetProperty("index").GetInt32(), got.Index);
                Assert.Equal(want.GetProperty("match").GetString(), got.Match);
                Assert.Equal(want.GetProperty("code").GetString(), got.Code);
                index++;
            }
        }
    }

    [Fact]
    public void RouteDiffsMatchTheSharedVector()
    {
        using var vector = ReadVector("peer-egress-routes-v1.json");
        var diffCases = vector.RootElement.GetProperty("diffCases");
        Assert.True(diffCases.GetArrayLength() > 0, "routes vector carried no diff cases");

        foreach (var testCase in diffCases.EnumerateArray())
        {
            var name = testCase.GetProperty("name").GetString()!;
            var difference = PeerEgressRoutePlanner.Diff(
                RoutesOf(testCase.GetProperty("current")), RoutesOf(testCase.GetProperty("desired")));
            AssertRoutes($"{name}/remove", difference.Remove, testCase.GetProperty("expect").GetProperty("remove"));
            AssertRoutes($"{name}/add", difference.Add, testCase.GetProperty("expect").GetProperty("add"));
        }
    }

    /// <summary>
    /// The parsers read output formats nobody controls, which makes them the part most likely to be
    /// wrong. Each runtime writing its own fixtures from its own reading of the man page is how
    /// three readings of one format come about, so all three read these.
    /// </summary>
    [Fact]
    public void RouteCommandParsingMatchesTheSharedVector()
    {
        using var vector = ReadVector("peer-egress-routes-v1.json");
        var commands = vector.RootElement.GetProperty("routeCommands");

        var routeGet = commands.GetProperty("routeGet");
        Assert.True(routeGet.GetArrayLength() > 0, "routes vector carried no routeGet cases");
        foreach (var testCase in routeGet.EnumerateArray())
        {
            var name = testCase.GetProperty("name").GetString()!;
            var hop = PeerEgressRouteCommands.ParseRouteGet(testCase.GetProperty("output").GetString());
            var expect = testCase.GetProperty("expect");
            if (!expect.GetProperty("parsed").GetBoolean())
            {
                Assert.True(hop is null, name);
                continue;
            }
            Assert.True(hop is not null, name);
            Assert.Equal(expect.GetProperty("gateway").GetString(), hop!.Value.Gateway);
            Assert.Equal(expect.GetProperty("device").GetString(), hop.Value.Device);
        }

        var showExact = commands.GetProperty("showExact");
        Assert.True(showExact.GetArrayLength() > 0, "routes vector carried no showExact cases");
        foreach (var testCase in showExact.EnumerateArray())
        {
            var name = testCase.GetProperty("name").GetString()!;
            var existing = PeerEgressRouteCommands.ParseShowExact(testCase.GetProperty("output").GetString());
            Assert.True(testCase.GetProperty("expect").GetProperty("present").GetBoolean() == existing.Present, name);
            Assert.Equal(testCase.GetProperty("expect").GetProperty("description").GetString(), existing.Description);
        }

        var tunnelDevice = commands.GetProperty("tunnelDevice").GetProperty("cases");
        Assert.True(tunnelDevice.GetArrayLength() > 0, "routes vector carried no tunnelDevice cases");
        foreach (var testCase in tunnelDevice.EnumerateArray())
        {
            var got = PeerEgressRouteCommands.HopIsDevice(
                new PeerEgressRouteHop(string.Empty, testCase.GetProperty("device").GetString()!),
                testCase.GetProperty("tun").GetString());
            Assert.True(testCase.GetProperty("expect").GetBoolean() == got,
                $"{testCase.GetProperty("device").GetString()} against {testCase.GetProperty("tun").GetString()}");
        }
    }
}

File: implementations/java/client/src/test/java/com/theshuai/specusclient/peer/PeerEgressRouteVectorTests.java
package com.theshuai.specusclient.peer;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertTrue;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.theshuai.common.peeregress.PeerEgressRule;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;
import org.junit.jupiter.api.Test;

/**
 * Replays {@code peer-egress-routes-v1.json} against the Java route planner.
 *
 * <p>Unlike the TCP fixture, this one's expectations come from an independent reference planner in
 * {@code tools/protocol/generate_peer_egress_route_vectors.py} rather than from any implementation,
 * so agreeing with it is independent agreement rather than a shared mistake.
 *
 * <p>The operator-visible {@code origin} string is asserted too. Three runtimes that installed the
 * same prefixes but described them differently would still leave an operator unable to compare one
 * machine's log with another's.
 */
class PeerEgressRouteVectorTests {

    private static final ObjectMapper MAPPER = new ObjectMapper();

    private static JsonNode readVector(String name) throws IOException {
        Path directory = Path.of("").toAbsolutePath();
        for (int depth = 0; depth < 8 && directory != null; depth++) {
            Path candidate = directory.resolve("protocol").resolve("test-vectors").resolve(name);
            if (Files.exists(candidate)) {
                return MAPPER.readTree(Files.readString(candidate));
            }
            directory = directory.getParent();
        }
        throw new IOException("cannot locate " + name);
    }

    private static List<PeerEgressRoutePlanner.Route> routesOf(JsonNode array) {
        List<PeerEgressRoutePlanner.Route> routes = new ArrayList<>();
        for (JsonNode node : array) {
            routes.add(new PeerEgressRoutePlanner.Route(
                    node.path("cidr").asText(),
                    PeerEgressRoutePlanner.Kind.fromWireName(node.path("kind").asText()),
                    node.path("origin").asText()));
        }
        return routes;
    }

    private static void assertRoutes(String label, List<PeerEgressRoutePlanner.Route> got, JsonNode want) {
        List<PeerEgressRoutePlanner.Route> expected = routesOf(want);
        assertEquals(expected.size(), got.size(), label + ": route count, got " + got);
        for (int index = 0; index < expected.size(); index++) {
            assertEquals(expected.get(index), got.get(index), label + ": route " + index);
        }
    }

    @Test
    void routePlansMatchTheSharedVector() throws IOException {
        JsonNode vector = readVector("peer-egress-routes-v1.json");
        String meshCidr = vector.path("meshCidr").asText();
        assertFalse(vector.path("planCases").isEmpty(), "routes vector carried no plan cases");

        for (JsonNode node : vector.path("planCases")) {
            String name = node.path("name").asText();
            List<PeerEgressRule> rules = new ArrayList<>();
            for (JsonNode raw : node.path("rules")) {
                rules.add(MAPPER.treeToValue(raw, PeerEgressRule.class));
            }
            List<String> bypass = new ArrayList<>();
            for (JsonNode raw : node.path("bypass")) {
                bypass.add(raw.asText());
            }

            PeerEgressRoutePlanner.Plan plan = PeerEgressRoutePlanner.plan(rules, bypass, meshCidr);
            assertRoutes(name, plan.routes(), node.path("expect").path("routes"));

            // Asserted alongside the routes because a planner that silently dropped a bad rule
            // would produce exactly the same route list.
            JsonNode wantRefused = node.path("expect").path("refused");
            assertEquals(wantRefused.size(), plan.refused().size(), name + ": refusal count");
            for (int index = 0; index < wantRefused.size(); index++) {
                JsonNode want = wantRefused.get(index);
                PeerEgressRoutePlanner.Refusal got = plan.refused().get(index);
                assertEquals(want.path("index").asInt(), got.index(), name + ": refusal " + index + " index");
                assertEquals(want.path("match").asText(), got.match(), name + ": refusal " + index + " match");
                assertEquals(want.path("code").asText(), got.code(), name + ": refusal " + index + " code");
            }
        }
    }

    @Test
    void routeDiffsMatchTheSharedVector() throws IOException {
        JsonNode vector = readVector("peer-egress-routes-v1.json");
        assertFalse(vector.path("diffCases").isEmpty(), "routes vector carried no diff cases");

        for (JsonNode node : vector.path("diffCases")) {
            String name = node.path("name").asText();
            PeerEgressRoutePlanner.Difference difference = PeerEgressRoutePlanner.diff(
                    routesOf(node.path("current")), routesOf(node.path("desired")));
            assertRoutes(name + "/remove", difference.remove(), node.path("expect").path("remove"));
            assertRoutes(name + "/add", difference.add(), node.path("expect").path("add"));
        }
    }

    /**
     * The parsers read output formats nobody controls, which makes them the part most likely to be
     * wrong. Each runtime writing its own fixtures from its own reading of the man page is how
     * three readings of one format come about, so all three read these.
     */
    @Test
    void routeCommandParsingMatchesTheSharedVector() throws IOException {
        JsonNode commands = readVector("peer-egress-routes-v1.json").path("routeCommands");
        assertFalse(commands.path("routeGet").isEmpty(), "routes vector carried no routeGet cases");

        for (JsonNode node : commands.path("routeGet")) {
            String name = node.path("name").asText();
            PeerEgressRouteCommands.Hop hop =
                    PeerEgressRouteCommands.parseRouteGet(node.path("output").asText());
            JsonNode expect = node.path("expect");
            if (!expect.path("parsed").asBoolean()) {
                assertNull(hop, name);
                continue;
            }
            assertNotNull(hop, name);
            assertEquals(expect.path("gateway").asText(), hop.gateway(), name);
            assertEquals(expect.path("device").asText(), hop.device(), name);
        }

        assertFalse(commands.path("showExact").isEmpty(), "routes vector carried no showExact cases");
        for (JsonNode node : commands.path("showExact")) {
            String name = node.path("name").asText();
            PeerEgressRouteCommands.ExistingRoute existing =
                    PeerEgressRouteCommands.parseShowExact(node.path("output").asText());
            assertEquals(node.path("expect").path("present").asBoolean(), existing.present(), name);
            assertEquals(node.path("expect").path("description").asText(), existing.description(), name);
        }

        for (JsonNode node : commands.path("tunnelDevice").path("cases")) {
            boolean got = PeerEgressRouteCommands.hopIsDevice(
                    new PeerEgressRouteCommands.Hop("", node.path("device").asText()),
                    node.path("tun").asText());
            String label = node.path("device").asText() + " against " + node.path("tun").asText();
            assertEquals(node.path("expect").asBoolean(), got, label);
        }
        assertTrue(commands.path("tunnelDevice").path("cases").size() > 0,
                "routes vector carried no tunnelDevice cases");
    }
}
